### PR TITLE
Perceptual loss

### PR DESCRIPTION
**Pull request by sanchezvvictor**

### CODE REVIEW

File: mfai/torch/losses/perceptual.py [hunk 1]
```diff
@@ -0,0 +1,510 @@
...
+                channel_iterative_mode: (bool) - To compute the Perceptual Loss over channels of the input (default = False)
+                in_channels: (int) - Number of input channels for perceptual Loss - [Used only if channel_iterative_mode=False] (default = 1)
+                pre_trained: (bool) - To use a pre-trained or a random version of the VGG16 (default = False)
+                resize_input: (bool) - To adapt input size to ImageNet Dataset size (224x224) (default = False)
+                style_layer_ids: (list) - Ids of Style Layers used for Perceptual Loss (default = [])
+                feature_layer_ids: (list) - Ids of Feature Layers used for Perceptual Loss (default = [4,9,16,23,30])
+                alpha_style: (float) - Weight of Style Loss (default = 0)
+                alpha_feature (float) - Weight of Feature Loss (default = 1)
+         '''
+        super(PerceptualLoss, self).__init__()
```
Comment (colon3ltocard) [inline mfai/torch/losses/perceptual.py:39]: Consider using Python3 super ?

File: mfai/torch/losses/perceptual.py [hunk 2]
```diff
@@ -0,0 +1,510 @@
...
+        # Multi-scale mode
+        self.multi_scale = multi_scale
+        self.scaling_factor = [0]
+        if multi_scale :
+            for i in range(3):
+                self.scaling_factor.append(2**i)
+        
+        self.__set_network()
+
+    def __set_blocks(self):
```
Comment (colon3ltocard) [inline mfai/torch/losses/perceptual.py:74]: What is your intention with using double underscore name mangling ?
See this https://stackoverflow.com/questions/1301346/what-is-the-meaning-of-single-and-double-underscore-before-an-object-name

**Comment by colon3ltocard** [comment]
@sanchezvvictor Great work ! Could you add a short documentation in the README.md ? Also explaining how to use your loss in a real use case ?

File: mfai/torch/losses/perceptual.py [hunk 3]
```diff
@@ -0,0 +1,510 @@
...
+                 pre_trained: bool = False,
+                 resize_input: bool = False,
+                 style_layer_ids: list = [],
+                 style_block_ids: list = [],
+                 feature_layer_ids: list = [4,9,16,23,30],
+                 feature_block_ids: list = [0,1,2,3,4],
+                 alpha_style: float = 0,
+                 alpha_feature: float = 1,
+        ):
+        r''' Class that computes the Perceptual Loss based on selected Network.
```
Comment (colon3ltocard) [inline mfai/torch/losses/perceptual.py:25]: Maybe add a link to the research paper/blog post explaining the ideas behind this loss ?

File: mfai/torch/losses/perceptual.py (hunk 1 above)
Comment (sanchezvvictor) [reply mfai/torch/losses/perceptual.py:39]: Not sure to understand your comment. What do you mean by 'Python3 super' ?

File: mfai/torch/losses/perceptual.py (hunk 2 above)
Comment (sanchezvvictor) [reply mfai/torch/losses/perceptual.py:74]: I have misgendered '__' with'_' sorry. The idea was to make these functions private.

File: mfai/torch/losses/perceptual.py (hunk 1 above)
Comment (colon3ltocard) [reply mfai/torch/losses/perceptual.py:57]: https://www.digitalocean.com/community/tutorials/python-super

Comment (colon3ltocard) [reply mfai/torch/losses/perceptual.py:57]: https://powerfulpython.com/blog/main-difference-python-3/

**Review by colon3ltocard — DISMISSED** [review]

**Review by colon3ltocard — DISMISSED** [review]

File: README.md [hunk 4]
```diff
@@ -162,6 +162,75 @@ In addition to metrics available in [**torchmetrics**](https://lightning.ai/docs
...
 
+# Losses
+
+Pytorch already provide some Loss like Mean Squared Error (torch.nn.MSELoss) or Mean Absolute Error (torch.nn.L1Loss). Here we add two loss functions that focus on perceptual similarity of tensors.
+
+## Perceptual Loss
+
+It was introduced by Johnson et al. - Perceptual losses for real-time style transfer and super-resolution. (https://arxiv.org/pdf/1603.08155).  
+
+The [**PerceptualLoss**](mfai/torch/losses/perceptual.py#L28) class is a torch.nn.Module that allows to initialize a VGG-16 and compute directly the perceptual loss between a given input and target.
```
Comment (tourniert) [inline README.md:173]:
```suggestion
The [**PerceptualLoss**](mfai/torch/losses/perceptual.py#L28) class is a `torch.nn.Module` that allows to initialize a VGG-16 and compute directly the perceptual loss between a given input and target.
```

File: README.md [hunk 5]
```diff
@@ -162,6 +162,75 @@ In addition to metrics available in [**torchmetrics**](https://lightning.ai/docs
...
+  targets = torch.rand(25, 5, 128, 128)
+
+  # The features of the targets are computed and compared to the input features
+  perceptual_loss = perceptual_loss_class(targets)
+
+```
+
+## LPIPS
+
+The [**LPIPS**](mfai/torch/losses/perceptual.py#L28) class is a torch.nn.Module that computes the Learned Perceptual Image Patch Similarity metric. It is using the aforementionned PerceptualLoss class so it contains the same modes.
```
Comment (tourniert) [inline README.md:231]:
```suggestion
The [**LPIPS**](mfai/torch/losses/perceptual.py#L28) class is a `torch.nn.Module` that computes the Learned Perceptual Image Patch Similarity metric. It is using the aforementionned PerceptualLoss class so it contains the same modes.
```

File: mypy_script.sh
Comment (tourniert) [inline mypy_script.sh:1]: Why did you have to add this shell script?

**Review by tourniert — APPROVED** [review]
